cpp/src/pausable_thread_loop.cpp

```diff
@@ -37,27 +51,40 @@ PausableThreadLoop::~PausableThreadLoop() {
 
 bool PausableThreadLoop::is_running() const noexcept {
     std::lock_guard<std::mutex> lock(mutex_);
-    return !paused_;
+    return state_ != State::Paused;
+}
+
+void PausableThreadLoop::pause_nb() {
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        if (state_ == State::Running) {
+            state_ = State::Pausing;
+        }
+    }
+    cv_.notify_one();
 }
 
 void PausableThreadLoop::pause() {
-    std::lock_guard<std::mutex> lock(mutex_);
-    paused_ = true;
+    pause_nb();
+    // And wait for the loop to flip to paused state. Behaviour is
+    // undefined if someone else called resume/stop while we're in
+    // this function.
+    std::unique_lock lock(mutex_);
+    cv_.wait(lock, [this]() { return state_ == State::Paused; });
 }
 
 void PausableThreadLoop::resume() {
     {
         std::lock_guard<std::mutex> lock(mutex_);
-        paused_ = false;
+        state_ = State::Running;
```

Comment:
Should we check `Stopping` here? Ideally once we have started stopping, we should not be able to resume